### ISSUE TITLE
JavaScript fails to call Java static method in native image

### ISSUE THREAD

**Issue by michael-simons** [oracle/graal#5382]
Given

```java
import static java.lang.System.out;
import org.graalvm.polyglot.*;

public class HelloJS {
  // native-image --language:js HelloJS
  public static void main(String... args) {
     var program = """
           const LocalDate = Java.type('java.time.LocalDate')
           let result = JSON.parse('{"greeting": "Hello, Øredev"}')
           result = { ...result, "when": LocalDate.of(2022, 11, 9)}
        """;

     try (var context = Context.newBuilder().allowAllAccess(true).build()) {
        var result = context.eval(Source.create("js", program));
        out.println(
           result.getMember("greeting") + " on " + result.getMember("when")
        );
     }
  }
}
```

I can run (with `gu install js`done)

```
java HelloJS 
Hello, Øredev on 2022-11-09
```

with

```
java --version           
openjdk 17.0.5 2022-10-18
OpenJDK Runtime Environment GraalVM CE 22.3.0 (build 17.0.5+8-jvmci-22.3-b08)
OpenJDK 64-Bit Server VM GraalVM CE 22.3.0 (build 17.0.5+8-jvmci-22.3-b08, mixed mode, sharing)
```

I expect this to work in native, too, however, this is what I get:

```
➜  oeredev javac HelloJS.java 
➜  oeredev native-image --language:js HelloJS                                                                          
========================================================================================================================
GraalVM Native Image: Generating 'hellojs' (executable)...
========================================================================================================================
[1/7] Initializing...                                                                                    (4,0s @ 0,31GB)
 Version info: 'GraalVM 22.3.0 Java 17 CE'
 Java version info: '17.0.5+8-jvmci-22.3-b08'
 C compiler: cc (apple, arm64, 14.0.0)
 Garbage collector: Serial GC
 3 user-specific feature(s)
 - com.oracle.svm.truffle.TruffleBaseFeature: Provides base support for Truffle
 - com.oracle.svm.truffle.TruffleFeature: Enables compilation of Truffle ASTs to machine code
 - org.graalvm.home.HomeFinderFeature: Finds GraalVM paths and its version number
[2/7] Performing analysis...  [*******]                                                                 (23,1s @ 2,93GB)
  16.096 (94,23%) of 17.082 classes reachable
  27.280 (64,09%) of 42.565 fields reachable
  84.877 (67,90%) of 125.003 methods reachable
  12.478 ( 9,98%) of 125.003 methods included for runtime compilation
     202 classes,   102 fields, and 1.306 methods registered for reflection
      59 classes,    59 fields, and    53 methods registered for JNI access
       4 native libraries: -framework Foundation, dl, pthread, z
[3/7] Building universe...                                                                               (1,9s @ 3,83GB)
[4/7] Parsing methods...      [**]                                                                       (4,3s @ 1,22GB)
[5/7] Inlining methods...     [****]                                                                     (2,4s @ 2,13GB)
[6/7] Compiling methods...    [*****]                                                                   (24,1s @ 2,72GB)
[7/7] Creating image...                                                                                  (6,5s @ 4,16GB)
  50,77MB (35,99%) for code area:    69.749 compilation units
  88,94MB (63,05%) for image heap:  681.980 objects and 3.726 resources
   1,36MB ( 0,96%) for other data
 141,06MB in total
------------------------------------------------------------------------------------------------------------------------
Top 10 packages in code area:                               Top 10 object types in image heap:
   7,47MB com.oracle.truffle.js.builtins                      24,36MB byte[] for embedded resources
   2,69MB com.oracle.truffle.js.nodes.access                  15,82MB byte[] for code metadata
   2,31MB com.oracle.truffle.js.nodes.binary                   9,00MB byte[] for graph encodings
   1,75MB com.oracle.truffle.js.builtins.temporal              4,61MB byte[] for general heap data
   1,57MB com.oracle.truffle.polyglot                          4,60MB java.lang.Class
   1,09MB java.util                                            3,72MB java.lang.String
   1,04MB com.oracle.truffle.api.interop                       3,08MB byte[] for java.lang.String
   1,01MB com.oracle.truffle.api.strings                       2,04MB com.oracle.svm.graal.meta.SubstrateField
1000,22KB com.oracle.truffle.host                              1,91MB c.o.truffle.api.nodes.NodeClassImpl$NodeFieldData
 986,20KB com.oracle.truffle.js.runtime.builtins               1,73MB com.oracle.svm.graal.meta.SubstrateMethod
  29,51MB for 383 more packages                               17,30MB for 3783 more object types
------------------------------------------------------------------------------------------------------------------------
                        4,0s (5,4% of total time) in 52 GCs | Peak RSS: 6,59GB | CPU load: 5,81
------------------------------------------------------------------------------------------------------------------------
Produced artifacts:
 /Users/msimons/Projects/Misc/oeredev/hellojs (executable)
 /Users/msimons/Projects/Misc/oeredev/hellojs.build_artifacts.txt (txt)
========================================================================================================================
Finished generating 'hellojs' in 1m 14s.
➜  oeredev ./hellojs 
Exception in thread "main" ReferenceError: of is not defined
	at <js> :program(Unnamed:1:21-22)
	at org.graalvm.sdk/org.graalvm.polyglot.Context.eval(Context.java:399)
	at HelloJS.main(HelloJS.java:15)
```

I am on macOS 13, Arm64.

**Comment by fniephaus** [comment]
### TL;DR
If you save 
```json
[
  {
    "name": "java.time.LocalDate",
    "methods": [
      {
        "name": "of",
        "parameterTypes": [ "int", "int", "int" ]
      }
    ]
  }
]
```
 as `reflect-config.json` and use

```bash
native-image --language:js -H:ReflectionConfigurationFiles=reflect-config.json HelloJS
```
to build your app, it will run as expected:

```bash
$ ./hellojs 
Hello, Øredev on 2022-11-09
```

---

### Explanation
The example application embeds GraalVM's JavaScript runtime and the JavaScript code accesses the `java.time.LocalDate` type from the host Java. GraalVM Native Image is unable to determine that the `LocalDate` type is dynamically accessed at run-time. The reason for this is, as you might have guessed already, that the language interoperability protocol is implemented based on reflection, and interop with host Java is no exception. If you had used Espresso to access `LocalDate`, the problem wouldn't have happened. The solution I proposed informs the Native Image static analysis explicitly that `LocalDate.of(int, int, int)` needs to be [registered for reflection](https://www.graalvm.org/22.3/reference-manual/native-image/dynamic-features/Reflection/). This way, it can be called reflectively, for example when called from GraalVM languages.
